panoply/datasource.py

```diff
@@ -7,24 +10,72 @@ class DataSource(events.Emitter):
     options = None
 
     # data source constructor
-    def __init__(self, source, options = {}):
-        super(DataSource, self).__init__()
-        
-        self.source = source
+    def __init__(self, source, options = {}, events = {}):
+        super(DataSource, self).__init__(events)
+
+        self.source = dict(source)
         self.options = options
 
+        if 'destination' not in source:
+            self.source['destination'] = source.get('type')
+
     # log a message
     def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
```

Comment:
I will change to docstring and add description for each argument.

*Re 'the server'*, I did write "controlled by this oauth" trying to explain that I was referring to the DS server

*Re `invalidate_token`*, it was actually `refresh_token` at first but since this function doesn't refresh the token every time and just invalidates it (make sure that it's valid and refreshes it only if it's not), it made more sense to change it.

*Re what this function does*, modifying the source is not a direct purpose of this function

*Re the `keys` argument*, I'm not sure what you mean by "fragile", but i'll change it